File: common/scene_context_schema.py
"""scene_context 统一 schema 定义。

字段模板与常量由 `config/schema.yaml` 提供（C# 与 Python 单一事实来源）。
"""

from __future__ import annotations

from .contracts import (
    FORCED_MOVEMENT_CONTEXT_KEY,
    RAID_BUFF_WINDOW_CONTEXT_KEY,
    TARGETABLE_WINDOW_CONTEXT_KEY,
    TARGET_COUNT_WINDOW_CONTEXT_KEY,
)
from .schema_config import load_schema_config

_schema = load_schema_config()
_scene = _schema["scene_context"]
_window_extra_fields = _scene["window_extra_fields"]

SCENE_CONTEXT_KEYS = (
    TARGETABLE_WINDOW_CONTEXT_KEY,
    FORCED_MOVEMENT_CONTEXT_KEY,
    RAID_BUFF_WINDOW_CONTEXT_KEY,
    TARGET_COUNT_WINDOW_CONTEXT_KEY,
)

TARGETABLE_SEGMENT_KINDS = tuple(_scene["targetable_segment_kinds"])
RAID_BUFF_WINDOW_SOURCE_KEYS = tuple(_scene["raid_buff_window_source_keys"])


def _window_feature_keys(
    extra_fields: tuple[str, ...],
    *,
    source_keys: tuple[str, ...] | None = None,
) -> tuple[str, ...]:
    """按窗口的时间字段 + 额外字段模板构造 feature keys。

    YAML 模板中的 `{key}` 占位由 source_keys 逐个展开（团辅窗口）。
    """
    expanded: list[str] = []
    for field in extra_fields:
        if "{key}" in field:
            for source_key in (source_keys or ()):
                expanded.append(field.format(key=source_key))
        else:
            expanded.append(field)
    return (
        *tuple(_scene["window_time_fields"]),
        *expanded,
    )


def raid_buff_window_feature_keys(source_keys: tuple[str, ...]) -> tuple[str, ...]:
    """按配置中的团辅标记技能构造 scene token feature keys。"""
    if not source_keys:
        raise ValueError("raid buff window source keys must not be empty")
    return _window_feature_keys(
        tuple(_window_extra_fields["raid_buff"]),
        source_keys=source_keys,
    )


TARGETABLE_WINDOW_FEATURE_KEYS = _window_feature_keys(
    (
        *tuple(_window_extra_fields["targetable"]),
        *(f"segment_kind.{kind}" for kind in TARGETABLE_SEGMENT_KINDS),
    )
)

FORCED_MOVEMENT_FEATURE_KEYS = _window_feature_keys(
    tuple(_window_extra_fields["forced_movement"])
)

RAID_BUFF_WINDOW_FEATURE_KEYS = raid_buff_window_feature_keys(RAID_BUFF_WINDOW_SOURCE_KEYS)

TARGET_COUNT_WINDOW_FEATURE_KEYS = _window_feature_keys(
    tuple(_window_extra_fields["target_count"])
)


def build_window_context(
    feature_keys: tuple[str, ...],
    tokens: list[list[float]] | None = None,
) -> dict[str, object]:
    """构造统一的窗口上下文结构。"""
    return {
        "feature_keys": list(feature_keys),
        "tokens": [] if tokens is None else tokens,
    }


def build_empty_scene_context() -> dict[str, object]:
    """构造稳定空壳 scene_context。"""
    return {
        TARGETABLE_WINDOW_CONTEXT_KEY: build_window_context(TARGETABLE_WINDOW_FEATURE_KEYS),
        FORCED_MOVEMENT_CONTEXT_KEY: build_window_context(FORCED_MOVEMENT_FEATURE_KEYS),
        RAID_BUFF_WINDOW_CONTEXT_KEY: build_window_context(RAID_BUFF_WINDOW_FEATURE_KEYS),
        TARGET_COUNT_WINDOW_CONTEXT_KEY: build_window_context(TARGET_COUNT_WINDOW_FEATURE_KEYS),
    }

File: Combat.Sim/FightEngine/Outputs/SceneContextSchema.cs
// Copyright (C) 2026 ICE_crystal
// SPDX-License-Identifier: GPL-3.0-only
// Additional permission: FightEngine GPLv3 Linking Exception, Version 1.0.
// See LICENSE and LICENSE-FightEngine-Linking-Exception in the repository root.

using Combat.Sim.Common;
using Combat.Sim.Config;

namespace Combat.Sim.Outputs;

/// <summary>
/// scene_context 统一 schema 定义（对照 outputs/scene_context_schema.py）。
/// 字段模板与常量由 <c>config/schema.yaml</c> 提供（C# 与 Python 单一事实来源），
/// 使用前必须先调用 <see cref="SchemaConfigLoader.Load"/>。
/// </summary>
public static class SceneContextSchema
{
    public static string[] SceneContextKeys =>
        new[]
        {
            SceneContracts.TargetableWindowContextKey,
            SceneContracts.ForcedMovementContextKey,
            SceneContracts.RaidBuffWindowContextKey,
            SceneContracts.TargetCountWindowContextKey,
        };

    public static string[] TargetableSegmentKinds =>
        SchemaConfigLoader.Instance.TargetableSegmentKinds.ToArray();

    public static string[] RaidBuffWindowSourceKeys =>
        SchemaConfigLoader.Instance.RaidBuffWindowSourceKeys.ToArray();

    private static string[] WindowTimeFields =>
        SchemaConfigLoader.Instance.WindowTimeFields.ToArray();

    private static string[] WindowExtraFields(string windowKey) =>
        SchemaConfigLoader.Instance.WindowExtraFields.TryGetValue(windowKey, out var fields)
            ? fields.ToArray()
            : throw new InvalidOperationException(
                $"schema.yaml 缺少 scene_context.window_extra_fields.{windowKey}");

    /// <summary>按配置中的团辅标记技能构造 scene token feature keys（对照 raid_buff_window_feature_keys）。</summary>
    public static string[] RaidBuffWindowFeatureKeysFor(IReadOnlyList<string> sourceKeys)
    {
        if (sourceKeys.Count == 0)
        {
            throw new ArgumentException("raid buff window source keys must not be empty", nameof(sourceKeys));
        }

        return WindowTimeFields
            .Concat(ExpandWindowExtraFields(WindowExtraFields("raid_buff"), sourceKeys))
            .ToArray();
    }

    /// <summary>
    /// 按窗口额外字段模板展开 feature keys；`{key}` 占位由 source_keys 逐个填充，
    /// 其余字段原样透出（对照 Python _window_feature_keys 的 field.format(key=...)）。
    /// </summary>
    private static IEnumerable<string> ExpandWindowExtraFields(
        IReadOnlyList<string> extraFields,
        IReadOnlyList<string> sourceKeys) =>
        extraFields.SelectMany(field =>
            field.Contains("{key}")
                ? sourceKeys.Select(sourceKey => field.Replace("{key}", sourceKey))
                : new[] { field });

    public static string[] TargetableWindowFeatureKeys =>
        WindowTimeFields
            .Concat(WindowExtraFields("targetable"))
            .Concat(TargetableSegmentKinds.Select(kind => $"segment_kind.{kind}"))
            .ToArray();

    public static string[] ForcedMovementWindowFeatureKeys =>
        WindowTimeFields
            .Concat(WindowExtraFields("forced_movement"))
            .ToArray();

    public static string[] RaidBuffWindowFeatureKeys =>
        RaidBuffWindowFeatureKeysFor(RaidBuffWindowSourceKeys);

    public static string[] TargetCountWindowFeatureKeys =>
        WindowTimeFields
            .Concat(WindowExtraFields("target_count"))
            .ToArray();

    /// <summary>构造统一的窗口上下文结构（对照 build_window_context）。</summary>
    public static Dictionary<string, object?> BuildWindowContext(
        IReadOnlyList<string> featureKeys,
        List<List<double>>? tokens = null) =>
        new()
        {
            ["feature_keys"] = featureKeys.ToList(),
            ["tokens"] = tokens ?? new List<List<double>>(),
        };

    /// <summary>构造稳定空壳 scene_context（对照 build_empty_scene_context）。</summary>
    public static Dictionary<string, object?> BuildEmptySceneContext() =>
        new()
        {
            [SceneContracts.TargetableWindowContextKey] = BuildWindowContext(TargetableWindowFeatureKeys),
            [SceneContracts.ForcedMovementContextKey] = BuildWindowContext(ForcedMovementWindowFeatureKeys),
            [SceneContracts.RaidBuffWindowContextKey] = BuildWindowContext(RaidBuffWindowFeatureKeys),
            [SceneContracts.TargetCountWindowContextKey] = BuildWindowContext(TargetCountWindowFeatureKeys),
        };
}
